sqlx-macros/src/derives/decode.rs

```diff
@@ -0,0 +1,210 @@
+use super::attributes::{
+    check_strong_enum_attributes, check_struct_attributes, check_transparent_attributes,
+    check_weak_enum_attributes, parse_attributes,
+};
+use quote::quote;
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+use syn::{
+    parse_quote, Arm, Data, DataEnum, DataStruct, DeriveInput, Field, Fields, FieldsNamed,
+    FieldsUnnamed, Stmt, Variant,
+};
+
+pub fn expand_derive_decode(input: &DeriveInput) -> syn::Result<proc_macro2::TokenStream> {
+    let attrs = parse_attributes(&input.attrs)?;
+    match &input.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Unnamed(FieldsUnnamed { unnamed, .. }),
+            ..
+        }) if unnamed.len() == 1 => {
+            expand_derive_decode_transparent(input, unnamed.first().unwrap())
+        }
+        Data::Enum(DataEnum { variants, .. }) => match attrs.repr {
+            Some(_) => expand_derive_decode_weak_enum(input, variants),
+            None => expand_derive_decode_strong_enum(input, variants),
+        },
+        Data::Struct(DataStruct {
+            fields: Fields::Named(FieldsNamed { named, .. }),
+            ..
+        }) => expand_derive_decode_struct(input, named),
+        Data::Union(_) => Err(syn::Error::new_spanned(input, "unions are not supported")),
+        Data::Struct(DataStruct {
+            fields: Fields::Unnamed(..),
+            ..
+        }) => Err(syn::Error::new_spanned(
+            input,
+            "structs with zero or more than one unnamed field are not supported",
+        )),
+        Data::Struct(DataStruct {
```

Comment:
I would prefer to sort all the `Data::Struct` arms together, then `Data::Enum` and then `Data::Union`.